In [1]:
import csv

# set floatpoint precision for printing to 2
%precision 2

with open('mpg.csv') as csvfile:
    mpg = list(csv.DictReader(csvfile))

# view first 3 elements of list created from csv file
mpg[:3]

[{'': '1',
  'class': 'compact',
  'cty': '18',
  'cyl': '4',
  'displ': '1.8',
  'drv': 'f',
  'fl': 'p',
  'hwy': '29',
  'manufacturer': 'audi',
  'model': 'a4',
  'trans': 'auto(l5)',
  'year': '1999'},
 {'': '2',
  'class': 'compact',
  'cty': '21',
  'cyl': '4',
  'displ': '1.8',
  'drv': 'f',
  'fl': 'p',
  'hwy': '29',
  'manufacturer': 'audi',
  'model': 'a4',
  'trans': 'manual(m5)',
  'year': '1999'},
 {'': '3',
  'class': 'compact',
  'cty': '20',
  'cyl': '4',
  'displ': '2',
  'drv': 'f',
  'fl': 'p',
  'hwy': '31',
  'manufacturer': 'audi',
  'model': 'a4',
  'trans': 'manual(m6)',
  'year': '2008'}]

Here we can see each element of this list is a car in a dict form, the keys for the dict corresponds to a column in the csv file.

Lets see how many dicts - or cars - we have in our dataset

In [2]:
len(mpg)

234

We have a dictionary for each of the 234 cars in the dataset.

To extract just the column names, or the keys in each of these dictionaries we can use the `keys()` function on the first element in the mpg dataset since we are assuming each element has the same keys or _row names_

In [9]:
mpg[0].keys()

dict_keys(['', 'class', 'model', 'fl', 'cty', 'displ', 'drv', 'hwy', 'cyl', 'manufacturer', 'year', 'trans'])

Here we can see we have class, model, fuel type, cty, mpg, engine volume, front or rear wheel drive, highway mpg, manufacturer, model, year, and transmission type.

**NOTE:** I happen to know this from working with the same __cars__ dataset in R. This can easily be found on google when looking at the values is not helping.

Now let's try and find the average city MPG across all cars in our CSV file.
To do this we want to sum the value for the 'cty' key for each car in the set and divide by the number of cars used (which we already know the set to contain 234 cars).

So that should look something like:  
sum (each value of cty in dataset mpg) / size (mpg)

In [12]:
sum(float(d['cty']) for d in mpg) / len(mpg)

16.86

Now let's do the same thing for average hwy mpg across all cars in the dataset

In [13]:
sum(float(d['hwy']) for d in mpg) / len(mpg)

23.44

Note that the average mpg for highway is significantly better than for city. This makes sense, as anyone who drives knows they get much better gas milage on the highway that they do in the city.